crates/ty_python_semantic/resources/mdtest/boolean/short_circuit.md

```diff
@@ -65,3 +65,132 @@ def _(flag1: bool, flag2: bool):
         # error: [possibly-unresolved-reference]
         reveal_type(z)  # revealed: Literal[1]
 ```
+
+## Inside if-else blocks, we can sometimes know that short-circuit couldn't happen
+
+When if-test contains `And` condition, in the scope of if-body we can be sure that the test is
+truthy and therefore short-circuiting couldn't happen. Similarly, when if-test contains `Or`
+condition, in the scope of if-else we can be sure that the test is falsy, and therefore
+short-circuiting couldn't happen.
+
+### And
+
+```py
+def _(flag: bool, number: int):
+    if flag and (x := number):
+        # x must be defined here
+        reveal_type(x)  # revealed: int & ~AlwaysFalsy
+    else:
+        # error: [possibly-unresolved-reference]
+        reveal_type(x)  # revealed: int
+
+    # error: [possibly-unresolved-reference]
+    reveal_type(x)  # revealed: int
+```
+
+### Or
+
+```py
+def _(flag: bool, number: int):
+    if flag or (x := number):
+        # error: [possibly-unresolved-reference]
+        reveal_type(x)  # revealed: int
```

Comment:
Similarly here, if `x` is defined it must be truthy, so this _could_ be `int & AlwaysTruthy` (but I don't think it's important that we implement that right now, it can just be a comment)